File: Strivers-A2Z-DSA-Sheet/12. Binary Search Trees/2. Practice Problems/python/12. Largest BST in Binary Tree.py
"""
Question:
Given a binary tree. Find the size of its largest subtree that is a Binary Search Tree.

Explanation:
- We use a recursive function to traverse the binary tree in a bottom-up manner.
- At each node, we check if the left and right subtrees are binary search trees.
- We return a triplet containing the size of the subtree, the minimum value, and the maximum value.

Time Complexity: O(n)
Space Complexity: O(h)
"""
from typing import Optional, Tuple

class Node:
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

def solve(root: Optional[Node]) -> Tuple[int, int, int]:
    if not root:
        return (0, float('inf'), float('-inf'))
    lef = solve(root.left)
    rig = solve(root.right)
    if lef[2] < root.data < rig[1]:
        return (lef[0] + rig[0] + 1, min(lef[1], root.data), max(rig[2], root.data))
    # here we are returning such values so that bst condition doesn't satisfy upwards
    return (max(lef[0], rig[0]), float('-inf'), float('inf'))

def largestBst(root: Optional[Node]) -> int:
    ans = solve(root)
    return ans[0]

File: Strivers-A2Z-DSA-Sheet/12. Binary Search Trees/2. Practice Problems/javascript/12. Largest BST in Binary Tree.js
/*
Question:
Given a binary tree. Find the size of its largest subtree that is a Binary Search Tree.

Explanation:
- We use a recursive function to traverse the binary tree in a bottom-up manner.
- At each node, we check if the left and right subtrees are binary search trees.
- We return a triplet containing the size of the subtree, the minimum value, and the maximum value.

Time Complexity: O(n)
Space Complexity: O(h)
*/

class Node {
    constructor(data, left = null, right = null) {
        this.data = data;
        this.left = left;
        this.right = right;
    }
}

function solve(root) {
    if (!root) return { size: 0, min: Infinity, max: -Infinity };
    const lef = solve(root.left);
    const rig = solve(root.right);
    if (lef.max < root.data && root.data < rig.min) {
        return {
            size: lef.size + rig.size + 1,
            min: Math.min(lef.min, root.data),
            max: Math.max(rig.max, root.data)
        };
    }
    // here we are returning such values so that bst condition doesn't satisfy upwards
    return { size: Math.max(lef.size, rig.size), min: -Infinity, max: Infinity };
}

function largestBst(root) {
    const ans = solve(root);
    return ans.size;
}
